### ISSUE TITLE
Please specify the use of CURRENT_TIMESTAMP to me 

### ISSUE THREAD

**Issue by kukrik** [qcubed/qcubed#1269]
I apologize in advance if my question can be seen silly. I've been looking for a solution for a long time, but I did not find a good answer.

1. But if I want to use the`  
`date_created` timestamp DEFAULT CURRENT_TIMESTAMP`
, it does not work with QCubed 3.0. This is needed, for example, to fix a new document, a new user for the first time, etc.

2.  This example, when using only the ` 
`date_updated` timestamp DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP
`, it works with QCubed 3.0.

Is there any solution to this question?

I would be very grateful if you find time to guide me.

**Comment by matthiaz** [comment]
Be careful using a TIMESTAMP column and DEFAULT_CURRENT_TIMESTAMP in QCubed.

The codegen will generate optimize locking constraint checks when you do.
This is useful but not if you don't expect/understand it. (see:
https://github.com/qcubed/qcubed/blob/master/includes/codegen/templates/db_orm/class_gen/object_save.tpl.php#L83
until
https://github.com/qcubed/qcubed/blob/master/includes/codegen/templates/db_orm/class_gen/object_save.tpl.php#L102
 )

The idea is that you protect your data from accidentil overwrites.
John and Peter both open a QForm with a User object on it
John changes the name of the user.
Peter changes the address of the user.
Both click the save button at roughly the same time.
John's request saves the user name and updates the timestamp column.
Peter's request will fail, because he is now trying to update the address
of the same user, but the last_modified timestamp he has on record, doesn't
match the one in the database.

If you don't want this behaviour:
1 You can modify the gens to not generate this code for any object
2 Or overload the save method in the object.
3 Or don't use a timestamp column.

I tend to use 3. I use regular datetime columns to store stuff like
date_created and date_updated and make sure that these dates get updated
using the Save method in the object. As long as you always use the Save
method on your objects to Insert/Update data, you are fine.

2017-07-02 17:08 GMT+02:00 Tiit Papp <notifications@github.com>:

**Comment by spekary** [comment]
What do you mean by "It does not work"? 

As Matthiaz pointed out, when you have an On Update set CurrentTime kind of trigger in your database, QCubed will use that to generate optimistic locking code.  In the past, the Optimistic Locking code could be problematic, and popup at strange times. For example, if you had a submit button on an edit form, and pressed it twice quickly, you might get an optimistic locking failure. That has been fixed for the most part in 3.1, and even better in 3.2. I added the ability to prevent a double-click from affecting a submit button ( a fourth parameter to the action), and the generated forms now put up a dialog if an optimistic locking exception is thrown, asking the user to confirm the save. In 3.2 there is a "OnClick" function for controls, which puts the debounce parameter in by default.

But I think your problem is different. You said something does not work when you are just trying to get a creation timestamp. My guess is you are running into model initialization problems.

v2 and v3.0 created an optional "initialize" method. But that initialize did not get called by default, you had to explicitly tell it to initialize by overriding the constructor of the model in your model subclass. I have changed that behavior for 3.1 so that the models are initialized by default based on the DEFAULT value in the SQL. 

Initialization is potentially impacted by the type of database you are using as well, since the code generator tries to figure out if you are trying to get a currrent timestamp, and will mirror that in the initialize function by setting a datetime to NOW. Otherwise, if you don't use the initialize function, it will set a datetime to NULL. I suspect that is what is happening to you.

**Comment by kukrik** [comment]
In any case, I'll let you know that I'm using  today QCubed 3.0.8.

@spekary , you correctly understood that it does not work, that is, it will set a datetime to NULL. In fact, I made the timestamp of the database, not datetime. Maybe the screenshot better explains.
![timestamp_example](https://user-images.githubusercontent.com/2853034/27842298-88225e64-6111-11e7-8074-dcf2679a33a2.jpg)
I hoped this opportunity was there. If not, then obviously I need to hear the @matthiaz's suggestion.

@spekary , or is it a possibility already exists? I would be very grateful, then I know better how to proceed.

**Comment by spekary** [comment]
Post the initialize() function from the generated model class for this table.

**Comment by kukrik** [comment]
```php
public function Initialize()
		{
			$this->intId = User::IdDefault;
			$this->strFirstName = User::FirstNameDefault;
			$this->strLastName = User::LastNameDefault;
			$this->strFullName = User::FullNameDefault;
			$this->strAvatar = User::AvatarDefault;
			$this->strEmail = User::EmailDefault;
			$this->strUsername = User::UsernameDefault;
			$this->strPassword = User::PasswordDefault;
			$this->blnPasswordResetFlag = User::PasswordResetFlagDefault;
			$this->intRoleTypeId = User::RoleTypeIdDefault;
			$this->intDefaultLanguageId = User::DefaultLanguageIdDefault;
			$this->intCountryId = User::CountryIdDefault;
			$this->blnStatus = User::StatusDefault;
			$this->dttDateCreated = (User::DateCreatedDefault === null)?null:new QDateTime(User::DateCreatedDefault);
			$this->strDateUpdated = User::DateUpdatedDefault;
		}
```
I understood, but what do you recommend to go about?

**Comment by spekary** [comment]
One potential way to fix this is this in your model override file for User:
```
const DateCreateDefault = QDateTime::Now;
```
Also, make sure that your constructor in User is uncommented and has this in it:
```
$this->Initialize();
```

That way, every time you create a User, it will automatically make a dttDateCreated value. Provided you save it right away, it will roughly correspond to your creation date and time in the database. You also might be running into the limitation that MySQL 5.6 and lower only allow one date field to be initialized with CURRENT_TIMESTAMP, and this will fix that problem too.

**Comment by matthiaz** [comment]
Why do it in the initialize?

Couldn't you do something like
if(is_null($this->dttDateCreated)) {
$this->dttDateCreated=QDateTime::Now(); } in the save method.?

2017-07-06 7:57 GMT+02:00 Shannon <notifications@github.com>:

**Comment by spekary** [comment]
That could work too. It depends a bit on the application. Usually, a new record and a retrieved record are displayed in the same Form/Panel. If you are displaying the creation date, do you want to show an empty value, or today's date there for a new record? I suppose that is a matter of taste for the developer, but its a little easier to just display the creation date, knowing it has a value, vs. checking to see if its null and displaying a blank if so. Not a big deal, just something to be aware of. If you try to initialize a QDateTime with a null, then display it, you get something like Jan. 1, 2000.

**Comment by kukrik** [comment]
@spekary, I've tried your recommendation, but it did not work for me. Perhaps my MySQL version is higher - 5.7.18? If I understand correctly?

If @matthiaz gave another recommendation and you did not consider the use of `is_null` to be useful. I think the @matthiaz recommendation is not bad if "traffic" is not active. I have not tried it yet, but I offer one option:
```php
if($blnIsNew) {
	$this->pnlUser->dttDateCreated  = QDateTime::Now();
} else {
	$this->pnlUser->dttDateUpdated  = QDateTime::Now();
};
```
Is not that a bad idea?

PS. In the database I changed the timestamp to datetime.
